File: src/enforcement/context.ts
/**
 * Author: Deepankar Das
 */

/**
 * Enforcer — Enforcement Context Builder
 *
 * Reads workspace path, repo name, branch, environment tier, and
 * deployment mode from environment variables and git metadata.
 * Attached to every ActionRequest.
 *
 * TDD Reference: Implementation Plan A8
 */

import * as fs from "node:fs";
import * as path from "node:path";
import { execSync } from "node:child_process";
import type { Environment } from "../../types/action.js";

/**
 * Detect git repo name from the .git directory or remote URL.
 */
function detectRepo(workspace: string): string {
  try {
    const remote = execSync("git remote get-url origin", {
      cwd: workspace,
      encoding: "utf-8",
      timeout: 3000,
    }).trim();
    // Extract org/repo from git URL
    const match = remote.match(/[:/]([^/]+\/[^/]+?)(?:\.git)?$/);
    return match ? match[1] : "";
  } catch {
    return "";
  }
}

/**
 * Detect current git branch.
 */
function detectBranch(workspace: string): string {
  try {
    return execSync("git branch --show-current", {
      cwd: workspace,
      encoding: "utf-8",
      timeout: 3000,
    }).trim();
  } catch {
    return "";
  }
}

/**
 * Detect deployment mode from environment.
 */
function detectDeploymentMode(): string {
  // Check if running inside Docker
  if (fs.existsSync("/.dockerenv")) {
    return "container";
  }
  // Check for remote workspace markers
  if (process.env.REMOTE_WORKSPACE || process.env.CODESPACE_NAME) {
    return "remote";
  }
  return "host";
}

/**
 * Build enforcement context from environment and git metadata.
 */
export function buildEnforcementContext(workspaceOverride?: string): Environment {
  const workspace = workspaceOverride || process.env.AA_WORKSPACE || process.cwd();
  const absoluteWorkspace = path.resolve(workspace);

  return {
    workspace: absoluteWorkspace,
    repo: process.env.AA_REPO || detectRepo(absoluteWorkspace),
    branch: process.env.AA_BRANCH || detectBranch(absoluteWorkspace),
    tier: process.env.AA_TIER || process.env.NODE_ENV || "development",
    deployment_mode: process.env.AA_DEPLOYMENT_MODE || detectDeploymentMode(),
  };
}

File: go/internal/enforcement/context.go
/**
 * Author: Deepankar Das
 */

package enforcement

import (
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"github.com/anthropics/enforcer/internal/types"
)

// BuildEnforcementContext creates an Environment from env vars and git metadata.
func BuildEnforcementContext(workspaceOverride string) types.Environment {
	workspace := envOrDefault("AA_WORKSPACE", workspaceOverride)
	if workspace == "" {
		workspace, _ = os.Getwd()
	}
	workspace, _ = filepath.Abs(workspace)

	return types.Environment{
		Workspace:      workspace,
		Repo:           envOrDefault("AA_REPO", detectRepo(workspace)),
		Branch:         envOrDefault("AA_BRANCH", detectBranch(workspace)),
		Tier:           envOrDefault("AA_TIER", envOrDefault("NODE_ENV", "development")),
		DeploymentMode: envOrDefault("AA_DEPLOYMENT_MODE", detectDeploymentMode()),
	}
}

func detectRepo(workspace string) string {
	cmd := exec.Command("git", "remote", "get-url", "origin")
	cmd.Dir = workspace
	out, err := cmd.Output()
	if err != nil {
		return "unknown/unknown"
	}
	remote := strings.TrimSpace(string(out))
	// Extract org/repo from URL
	remote = strings.TrimSuffix(remote, ".git")
	if idx := strings.LastIndex(remote, ":"); idx > 0 {
		return remote[idx+1:]
	}
	parts := strings.Split(remote, "/")
	if len(parts) >= 2 {
		return parts[len(parts)-2] + "/" + parts[len(parts)-1]
	}
	return remote
}

func detectBranch(workspace string) string {
	cmd := exec.Command("git", "branch", "--show-current")
	cmd.Dir = workspace
	out, err := cmd.Output()
	if err != nil {
		return "unknown"
	}
	return strings.TrimSpace(string(out))
}

func detectDeploymentMode() string {
	if _, err := os.Stat("/.dockerenv"); err == nil {
		return "container"
	}
	if os.Getenv("REMOTE_WORKSPACE") != "" || os.Getenv("CODESPACE_NAME") != "" {
		return "remote"
	}
	return "host"
}

func envOrDefault(key, fallback string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return fallback
}
